File: interpreter_python/parser.py
from lexer import Lexer, TokenType
from ast_nodes import Number, Boolean, BinOp, UnaryOp, Variable, VarAssign
from errors import ParserError

class Parser:
    def __init__(self, lexer: Lexer):
        self.lexer = lexer
        self.current_token = self.lexer.get_next_token()
        self.peek_token = self.lexer.get_next_token()

    def eat(self, token_type):
        if self.current_token.type == token_type:
            self.current_token = self.peek_token
            self.peek_token = self.lexer.get_next_token()
        else:
            raise ParserError(
                f"Invalid syntax: Expected {token_type}, got {self.current_token.type}",
                self.current_token.lineno,
                self.current_token.column
            )

    def primary(self):
        token = self.current_token
        
        if token.type == TokenType.NUMBER:
            self.eat(TokenType.NUMBER)
            return Number(token.value)
        
        if token.type == TokenType.TRUE:
            self.eat(TokenType.TRUE)
            return Boolean(True)
        
        if token.type == TokenType.FALSE:
            self.eat(TokenType.FALSE)
            return Boolean(False)
            
        if token.type == TokenType.LPAREN:
            self.eat(TokenType.LPAREN)
            node = self.expr()
            self.eat(TokenType.RPAREN)
            return node
            
        if token.type == TokenType.IDENTIFIER:
            # Check for assignment
            # For now, just variable access. Assignment usually handled at statement level.
            # But per grammar: assignment = identifier "=" expression
            # and expression is lower.
            # We'll handle assignment at top level if needed, strictly speaking grammar says 'statement'.
            # Here we assume inside an expression it's a variable usage.
            self.eat(TokenType.IDENTIFIER)
            return Variable(token.value)
            
        raise ParserError(f"Syntax Error: Unexpected token {token}", token.lineno, token.column)

    def unary(self):
        token = self.current_token
        if token.type in (TokenType.PLUS, TokenType.MINUS, TokenType.NOT):
            self.eat(token.type)
            return UnaryOp(op=token.value if token.value else token.type.name, operand=self.unary())
        return self.primary()

    def multiplicative(self):
        node = self.unary()
        while self.current_token.type in (TokenType.MUL, TokenType.DIV, TokenType.MOD):
            token = self.current_token
            self.eat(token.type)
            node = BinOp(left=node, op=token.value, right=self.unary())
        return node

    def additive(self):
        node = self.multiplicative()
        while self.current_token.type in (TokenType.PLUS, TokenType.MINUS):
            token = self.current_token
            self.eat(token.type)
            node = BinOp(left=node, op=token.value, right=self.multiplicative())
        return node

    def relational(self):
        node = self.additive()
        while self.current_token.type in (TokenType.LT, TokenType.GT, TokenType.LE, TokenType.GE):
            token = self.current_token
            self.eat(token.type)
            node = BinOp(left=node, op=token.value, right=self.additive())
        return node

    def equality(self):
        node = self.relational()
        while self.current_token.type in (TokenType.EQ, TokenType.NE):
            token = self.current_token
            self.eat(token.type)
            node = BinOp(left=node, op=token.value, right=self.relational())
        return node

    def logical_and(self):
        node = self.equality()
        while self.current_token.type == TokenType.AND:
            token = self.current_token
            self.eat(TokenType.AND)
            node = BinOp(left=node, op='and', right=self.equality())
        return node

    def logical_or(self):
        node = self.logical_and()
        while self.current_token.type == TokenType.OR:
            token = self.current_token
            self.eat(TokenType.OR)
            node = BinOp(left=node, op='or', right=self.logical_and())
        return node

    def expr(self):
        # Entry point for expressions
        return self.logical_or() 

    def parse(self):
        # Lookahead to distinguish assignment "x = ..." from expression "x + ..."
        if self.current_token.type == TokenType.IDENTIFIER and self.peek_token.type == TokenType.ASSIGN:
            var_name = self.current_token.value
            self.eat(TokenType.IDENTIFIER)
            self.eat(TokenType.ASSIGN)
            # Use expr() to parse the right-hand side value
            value = self.expr()
            return VarAssign(var_name, value)
            
        return self.expr()

File: interpreter_prolog/parser.pl
:- module(parser, [parse/2]).
:- use_module(lexer).

% Grammar Rules (DCG)
parse(Tokens, AST) :-
    expression(AST, Tokens, []).

% Hierarchy: logical_or -> logical_and -> equality -> relational -> additive -> multiplicative -> unary -> primary

% Hierarchy: logical_or -> logical_and -> equality -> relational -> additive -> multiplicative -> unary -> primary

expression(assign(Name, Expr)) --> [ident(Name)], [assign], expression(Expr).
expression(AST) --> logical_or(AST).

% Logical OR
logical_or(AST) --> logical_and(Left), logical_or_rest(Left, AST).

logical_or_rest(Acc, Final) --> 
    [or], logical_and(Right), 
    { NewAcc = binOp(or, Acc, Right) },
    logical_or_rest(NewAcc, Final).
logical_or_rest(Acc, Acc) --> [].

% Logical AND
logical_and(AST) --> equality(Left), logical_and_rest(Left, AST).

logical_and_rest(Acc, Final) --> 
    [and], equality(Right), 
    { NewAcc = binOp(and, Acc, Right) },
    logical_and_rest(NewAcc, Final).
logical_and_rest(Acc, Acc) --> [].

% Equality
equality(AST) --> relational(Left), equality_rest(Left, AST).

equality_rest(Acc, Final) --> 
    [eq], relational(Right), 
    { NewAcc = binOp(eq, Acc, Right) },
    equality_rest(NewAcc, Final).
equality_rest(Acc, Final) --> 
    [neq], relational(Right), 
    { NewAcc = binOp(neq, Acc, Right) },
    equality_rest(NewAcc, Final).
equality_rest(Acc, Acc) --> [].

% Relational
relational(AST) --> additive(Left), relational_rest(Left, AST).

relational_rest(Acc, Final) --> 
    [lt], additive(Right), 
    { NewAcc = binOp(lt, Acc, Right) },
    relational_rest(NewAcc, Final).
relational_rest(Acc, Final) --> 
    [gt], additive(Right), 
    { NewAcc = binOp(gt, Acc, Right) },
    relational_rest(NewAcc, Final).
relational_rest(Acc, Final) --> 
    [le], additive(Right), 
    { NewAcc = binOp(le, Acc, Right) },
    relational_rest(NewAcc, Final).
relational_rest(Acc, Final) --> 
    [ge], additive(Right), 
    { NewAcc = binOp(ge, Acc, Right) },
    relational_rest(NewAcc, Final).
relational_rest(Acc, Acc) --> [].

% Additive
additive(AST) --> multiplicative(Left), additive_rest(Left, AST).

additive_rest(Acc, Final) --> 
    [plus], multiplicative(Right), 
    { NewAcc = binOp(plus, Acc, Right) },
    additive_rest(NewAcc, Final).
additive_rest(Acc, Final) --> 
    [minus], multiplicative(Right), 
    { NewAcc = binOp(minus, Acc, Right) },
    additive_rest(NewAcc, Final).
additive_rest(Acc, Acc) --> [].

% Multiplicative
multiplicative(AST) --> unary(Left), multiplicative_rest(Left, AST).

multiplicative_rest(Acc, Final) --> 
    [times], unary(Right), 
    { NewAcc = binOp(times, Acc, Right) },
    multiplicative_rest(NewAcc, Final).
multiplicative_rest(Acc, Final) --> 
    [div], unary(Right), 
    { NewAcc = binOp(div, Acc, Right) },
    multiplicative_rest(NewAcc, Final).
multiplicative_rest(Acc, Final) --> 
    [mod], unary(Right), 
    { NewAcc = binOp(mod, Acc, Right) },
    multiplicative_rest(NewAcc, Final).
multiplicative_rest(Acc, Acc) --> [].

% Unary
unary(unaryOp(minus, Operand)) --> [minus], unary(Operand).
unary(unaryOp(not, Operand)) --> [not], unary(Operand).
unary(AST) --> primary(AST).

% Primary
primary(num(N)) --> [num(N)].
primary(bool(true)) --> [bool(true)].
primary(bool(false)) --> [bool(false)].
primary(Expr) --> [lparen], expression(Expr), [rparen].
primary(var(Name)) --> [ident(Name)].
